openshift/console
# Bug 2000607: Add an alert for the domain mapping linked with other knative service

**Fixes**: 

https://issues.redhat.com/browse/ODC-6305

**Analysis / Root cause**: 

Moving from the domain from one resource to another is not intuitive. The user should be given a warning or message.


**Solution Description**: 


1. Add an alert in the route section if  any one of the selected  domain mappings is already linked to a different knative service 
2. Add the knative service name next to the domain mapping as a dropdown item.
3. Added a extra validation so that newly created name should contain two segments seperated by dot.


**Screen shots / Gifs for design review**: 


GIF:

![mp-domain](https://user-images.githubusercontent.com/9964343/132002682-2962b45b-f3ea-441f-9c10-04b6bd099009.gif)


Alert for using a domain mapping from different knative service:
![image](https://user-images.githubusercontent.com/9964343/132002510-b6560aa7-149d-4beb-a2a7-22217f3e61b7.png)
Dropdown name with ksvc name :
![image](https://user-images.githubusercontent.com/9964343/131852217-8ec0109f-eb34-440f-87f5-1818f823f2de.png)


**Unit test coverage report**: 

**Serverless-utils:**
```
  hasOtherKsvcDomainMappings
    ✓ should return false if the domain mapping is empty
    ✓ should return false if the other ksvc info is not present in domain mapping
    ✓ should return true if the other ksvc info is present in domain mapping
  removeDuplicateDomainMappings
    ✓ should return empty arary if the domain mapping empty or null (1ms)
    ✓ should return unique value in the array (1ms)
    ✓ should strip the ksvc info and return unique value in the array
  getOtherKsvcFromDomainMapping
    ✓ should return null for invalid values
    ✓ should return null the ksvc name if is matching with the current ksvc name
    ✓ should return the ksvc name if its not matching with the current ksvc name (1ms)
  removeKsvcInfoFromDomainMapping
    ✓ should return empty string
    ✓ should return the same domain name if the ksvc info is not in the string
    ✓ should return only the domain name if the ksvc info is the string
   getAllOtherDomainMappingInUse
    ✓ should return empty array for the empty input
    ✓ should return empty array if selected domain is not already in the cluster (1ms)
    ✓ should return empty if the selected domain is mapped to current knative service
    ✓ should return domains if the selected domains are not mapped to current Knative service (1ms)
     ✓ should not return domains if the user creates domain with the same name again
```
**ServerlessRouteSection.spec.tsx:**
```
    ✓ Should render SelectInputField with the domain mapping options without other ksvc info (3ms)
    ✓ Should not contain the warning Alert if the selected domain is not from other knative services (3ms)
    ✓ Should contain the warning Alert if any of the selected domains is from other knative services (2ms)
```

**Test setup:**


1. Create a serverless workload with domain mapping
2. Check the domain mapping details in the workload sidebar
3. Create another workload using same domain mapping
4. Check the domain mapping details on the first workload

**Browser conformance**: 


cc: @invincibleJai
